travis: Run tests as part of travis CI

alistair23 opened the pull request:
Extend the travis script to actually run a libtock-rs app in QEMU.

To do this we have to drop OS X support.

Signed-off-by: Alistair Francis <alistair.francis@wdc.com>

torfmaster commented:
Wow, this is amazing! I imagine the integration tests running in qemu (I just wonder how to connect the gpio pins, though).

torfmaster commented:
bors try

torfmaster commented on .travis.yml:
```diff
@@ -20,6 +19,27 @@ install:
+  # Run a QEMU instance of the HiFive1 app
+  - PLATFORM=riscv32 cargo build --target=riscv32imac-unknown-none-elf --example hello_world
+  - elf2tab -o ./Tock-app.tab ./target/riscv32imac-unknown-none-elf/debug/examples/hello_world --stack 2048 --app-heap 1024 --kernel-heap 1024 -v --protected-region-size=64
+  - timeout --foreground 10s ./qemu-4.2.0/riscv32-softmmu/qemu-system-riscv32 -M sifive_e -kernel ../tock/boards/hifive1/target/riscv32imac-unknown-none-elf/release/hifive1 -device loader,file=./target/riscv32imac-unknown-none-elf/debug/examples/hello_world.tbf,addr=0x20430000 -nographic | tee serial
```
This script was intended to be run locally. I would leave the integration tests out for now as they involve compiling qemu.

alistair23 replied:
I'm not clear what you mean here

torfmaster replied:
Just a mistake, I thought these lines were in the run all checks script. Sorry.

torfmaster commented on .travis.yml:
```diff
@@ -20,6 +19,27 @@ install:
   - rustup component add clippy
+  - cargo install -f elf2tab --version 0.4.0
+  # Ubuntu doesn't package QEMU for RISC-V, so let's build it
+  - wget https://download.qemu.org/qemu-4.2.0.tar.xz
```
Is there a way to get artifacts for qemu somehow?

alistair23 replied:
Yep, it actually is packaged in Debian, I just couldn't find it.

torfmaster commented on .travis.yml:
```diff
@@ -20,6 +19,27 @@ install:
+  - git clone --depth=50 https://github.com/tock/tock.git
+  - cd tock/boards/hifive1
+  # Use a known working version of Tock
+  - git checkout c94059d3e25dc635e682facff4894ef43b9aca0e
```
Maybe we could package the kernel as an artifact inside the tock os build. Then we wouldn't have to rebuild it all the time.

alistair23 replied:
That would be cool! I can look into that.

jrvanwhy replied:
I have concerns about the `--depth=50` breaking quickly, though those are only relevant if you don't check in the kernel as a build artifact.

If you do check it in, it's probably worth providing the build script you used so we can reproduce the binary.

alistair23 replied:
It looks like it isn't simple to get the build artifacts from Tock, we would have to save it to a S3 instance: https://docs.travis-ci.com/user/uploading-artifacts/

I have removed the `--depth 50` as I agree.

torfmaster replied:
We could use the github releases flow to provide artifacts. However, for now (as tock releases are done manually) we can just attach kernel binaries to each release until there is an automated workflow. I think the tests should run ideally with the version provided in the Readme to really test compatibility. Now, it's not possible because as I understand there is no tock release supporting riscv userland apps.

alistair23 replied:
I agree it would be best to use a released version of Tock binaries, but currently that isn't an option. After the next release we can change to just downloading the release.

torfmaster replied:
:+1:

Woyten commented:
> This is great! I think this is worth dropping the mac os x build.

Yes. The build pipeline is becoming too slow. We could either remove the OS-X build or trigger a leight-weight `build_examples.sh`-only job without any testing on OS-X.

alistair23 commented on .travis.yml:
```diff
@@ -20,6 +19,27 @@ install:
+  - PLATFORM=riscv32 cargo build --target=riscv32imac-unknown-none-elf --example hello_world
+  - elf2tab -o ./Tock-app.tab ./target/riscv32imac-unknown-none-elf/debug/examples/hello_world --stack 2048 --app-heap 1024 --kernel-heap 1024 -v --protected-region-size=64
+  - timeout --foreground 10s ./qemu-4.2.0/riscv32-softmmu/qemu-system-riscv32 -M sifive_e -kernel ../tock/boards/hifive1/target/riscv32imac-unknown-none-elf/release/hifive1 -device loader,file=./target/riscv32imac-unknown-none-elf/debug/examples/hello_world.tbf,addr=0x20430000 -nographic | tee serial
+  - grep "Hello Tock World" serial
```
It might not run though, I have only managed to run `hello_world` on the HiFive1 due to space constraints.

jrvanwhy replied:
Asserting on the complete output would make the test fragile regardless; if we could run `libtock_test` then we would want to compute whether all test cases passed and use that result. It doesn't matter yet, though, as `libtock_test` is too large to run.

alistair23 commented on .travis.yml:
```diff
@@ -20,6 +19,27 @@ install:
   - rustup target add riscv32imc-unknown-none-elf
   - rustup component add rustfmt
   - rustup component add clippy
+  - cargo install -f elf2tab --version 0.4.0
```
I have removed it.

alistair23 commented:
I have pushed an update addressing all the comments.

Unfortunately qemu-system-riscv* isn't packaged in any of the current Ubuntu LTS releases so we still have to build QEMU. I have documented this in the Travis file and once the latest Ubuntu LTS is out and supported in Travis we can update to using that instead.

alistair23 commented on .travis.yml:
```diff
@@ -20,6 +19,27 @@ install:
   - rustup target add riscv32imc-unknown-none-elf
   - rustup component add rustfmt
   - rustup component add clippy
+  - cargo install -f elf2tab --version 0.4.0
```
I would rather not as I'm not a fan of having all these scripts around that install stuff on your computer.

Woyten commented:
Just saw that my review was in state "changes requested". Please don't hesitate to bors r+ nevertheless.

Woyten commented:
bors r+
